### PR TITLE
Generate valid structure for MakeCredential params

### PR DESCRIPTION
**Pull request by hcyang-google**
The fuzzers are not working well currently because the logic is to parse random bytes as cbor value and parse them as specific command format. If the format doesn't match, it returns and goes to the next round. With random inputs, this has ~0% probability of generating valid input, so we're only fuzzing the command parsing (ineffectively too, as the bytes can't be interpreted as cbor maps most of the time). We should generate valid inputs in each fuzzing round with high probability.

To do that, we need to generate valid cbor maps that can be parsed into specific parameters (like AuthenticatorMakeCredentialParameters). Here, I mostly take use of the arbitrary crate that is the default choice for Rust fuzzing. I decided to put the extra code in the ctap package itself behind a feature flag because otherwise we'll have to bring many extra dependency and make many types/functions public to implement it in the fuzz package. Even if we do that it's not enough, because we can't implement foreign traits on foreign types, so we'll likely have to rewrite our own arbitrary library as well. I feel that the way I add the fuzz-related code to the ctap library doesn't really complicates the code base so it should be fine. If you have better suggestions please tell me, let us first fix this CL to an acceptable format before we implement similar logic to other fuzz targets.

After this CL, we have 91.67% function coverage in ctap/hid/mod.rs (compared to 0%), and decently covers main logic in ctap/mod.rs for MakeCredential command processing. Main limitation of coverage is crypto-related checks, like we return an error if PubKeyCredentialsParams doesn't include ES256. These implementation-specific improvements can be added later.

Finally, I want to discuss our attitudes toward introducing probability in fuzzing. Theoretically as long as the inputs are reachable, we should let the fuzzer find its paths, and not choosing some specific paths for very high probability. But in practice we need to use probabilities or else fuzzing is very ineffective or even useless. I used 95% for valid structure, but for other cases (like whether PubKeyCredentialsParams include ES256), do we want to set higher probabilities for those "more interesting" paths too?

### CODE REVIEW

File: libraries/cbor/src/values.rs
Comment (hcyang-google) [inline libraries/cbor/src/values.rs:228]: Is there specific reasons we don't have this before? I need to auto parse `Vec<Value>` in cbor_map! macro. If we have `From<Vec<(Value, Value)>>` for map I think we can have this for array too.

Comment (kaczmarczyck) [reply libraries/cbor/src/values.rs:228]: My reasoning was that I'd rather have people explicitly choose what they want, so they don't create the wrong type accidentally. There are currently different types that rely on `Vec` in some form or another:

- ByteString (`Vec<u8>`)
- Array (`Vec<Value>`)
- Map (`Vec<(Value, Value>)`)

I'm not against it, if you think it improves the API, we should have it. Also I'm thinking about API changes to fix #420 , so we can also merge this as is and I'll consider your use case later.

File: Cargo.toml
Comment (kaczmarczyck) [inline Cargo.toml]: I noticed that we are using inconsistent string delimiters throughout our `.toml` files. We should probably only use " going forward?

File: fuzz/fuzz_helper/src/lib.rs
Comment (kaczmarczyck) [inline fuzz/fuzz_helper/src/lib.rs]: I assume this is still deterministic from a reproducibility point of view?

File: fuzz/fuzz_helper/src/lib.rs
Comment (kaczmarczyck) [inline fuzz/fuzz_helper/src/lib.rs]: Is this a rule of thumb, or did you hyperparameter tune this a bit, or just a guess?

File: src/ctap/mod.rs
Comment (kaczmarczyck) [inline src/ctap/mod.rs]: You could call the `cbor::writer` directly in fuzzing. This is the CTAP specific function with a CTAP error message. That has the advantage that fuzzing can play with different input nesting depths (that should still be parsed correctly, even when they don't match our expectatations).

So this doesn't have to be public then!

File: src/ctap/data_formats.rs
Comment (kaczmarczyck) [inline src/ctap/data_formats.rs:18]: I like the clear separation. Did you check binary size with the conditional compilation though? We realized that the compiler is smart enough to remove unused code. It might still make sense for readability.

@ia0 Opinions?

File: fuzz/fuzz_helper/src/lib.rs
Comment (hcyang-google) [reply fuzz/fuzz_helper/src/lib.rs]: Yes, deterministic.

File: fuzz/fuzz_helper/src/lib.rs
Comment (hcyang-google) [reply fuzz/fuzz_helper/src/lib.rs]: This is just a guess. I'm imagining something like HIGH_PROBABILITY, GOOD_PROBABILITY probs, to correspond to different levels of requirements we want. Like we "almost always" want the structure to be valid, and "prefer" the pub key cred params to include ES256. But I don't know good ways to determine their best values.

File: libraries/cbor/src/values.rs
Comment (ia0) [reply libraries/cbor/src/values.rs:228]: Yes, I agree those conversions should be unambiguous given the input and output types (i.e. there's only one way to go from the input to the output).

File: src/ctap/data_formats.rs
Comment (ia0) [reply src/ctap/data_formats.rs:18]: There's 2 things to consider:
1. Only guarding code that would otherwise to be statically eliminated.
2. Depending non-optionally from the `arbitrary` crate.

I think it makes sense to only depend optionally from `arbitrary`. This permits to make it clearer what are the real dependencies. And as a consequence of this, all the `derive(Arbitrary)` need to be guarded by the presence of the `arbitrary` crate otherwise they won't compile. So all in all, I'm fine with the current choice.

File: src/ctap/data_formats_fuzz.rs
Comment (ia0) [inline src/ctap/data_formats_fuzz.rs]: Our `rustfmt` doesn't enforce anything about [`imports_granularity`](https://rust-lang.github.io/rustfmt/?version=v1.4.38&search=#imports_granularity) but maybe we should decide something? I think in practice we use `Module` so I would vote in favor of continuing this practice and maybe even enforcing it in the `rustfmt` configuration.

@kaczmarczyck What do you think?

File: src/ctap/mod.rs
Comment (hcyang-google) [reply src/ctap/mod.rs]: Done

File: Cargo.toml
Comment (hcyang-google) [reply Cargo.toml]: Done

File: src/ctap/data_formats_fuzz.rs
Comment (ia0) [inline src/ctap/data_formats_fuzz.rs]: I don't understand why we need this trait compared to converting the output of `Arbitrary` on the given type to CBOR. (I'll try to discuss higher level questions like mutations somewhere else if I don't forget.)

File: fuzz/fuzz_helper/src/lib.rs
Comment (ia0) [inline fuzz/fuzz_helper/src/lib.rs]: We shouldn't use `data` here because it's biased. We should use `unstructured.take_rest()`.

**Review by ia0 — COMMENTED** [review]
This PR raises good questions with non-obvious answers. Here are some high-level comments:
- The PR description mentions "0% probability with random inputs", but fuzzing is not random, it's coverage-guided. For example, fuzzers are able to reconstruct file formats like PDF without help.
- The PR description mentions coverage. How was this measured? And on which corpus? I'd suggest using the existing OSS-Fuzz corpus or building one with 24h to 7d of continuous sequential fuzzing.
- One way to help fuzzing is to provide examples or a starting corpus.
- The problem of converting the fuzzer input to valid input only a fraction of the time means the fuzzer input is not homogeneous and thus can't be easily reused/mutated. A better approach would be to have 2 fuzzing targets: one using the fuzzer input as raw input, and one using the fuzzer input to generate valid input. In some way, the first version is more end-to-end while the second is more like fuzzing the internal functions (we could actually fuzz them directly to avoid unnecessary processing which could accidentally restrict coverage).
- Regarding validity, there's also the questions of mutation. Do we want to slightly mutate the input such that it may also be invalid although it looks valid (e.g. reorder items in a map).

I guess the ultimate goal of this PR is to improve fuzzing coverage, so for that I think a prerequisite is to agree on what is coverage and how we measure it. For example, for the persistent store fuzzing, the coverage is behavior based (in particular it's not about code) and is measured with the example binary when the argument is a directory. There's also an unsubmitted `libraries/persistent_store/coverage.sh` script in the `fuzz_coverage` branch of my fork that measures code coverage. Should eventually make a PR of it.

File: libraries/cbor/src/values.rs
Comment (kaczmarczyck) [reply libraries/cbor/src/values.rs:228]: Actually my list above was incorrectly formatted, I fixed it. It is not ambiguous, since there is no notion of tuples in CBOR, and u8 is different to Value. It might be tricky if we extend this to macros, but it's okay as is I think.
A user could try to call `.into()` on a vector of `Value`s of `u8` and expect it to become a ByteString, if they accidentally casted their array elements into `Value`. But since @hcyang-google found this more intuitive, we have real world evidence that this is better.

Comment (ia0) [reply libraries/cbor/src/values.rs:228]: Ah I see. If the `Value` version of a `Vec<Value>` is unambiguously an array (including in the future), then SGTM.

File: src/ctap/data_formats_fuzz.rs
Comment (kaczmarczyck) [reply src/ctap/data_formats_fuzz.rs]: Would have made it easier in this case to meet our style, just run `cargo fmt`. So if we care here, and I think we should, then we should merge #445 :)

**Comment by kaczmarczyck** [comment]
I did run
```
cargo fuzz run fuzz_target_process_ctap2_make_credential
```
with and without this PR, and the coverage numbers do look much better after a few minutes, at least. It might of course also make it harder to find obscure CTAP bugs. But we do have different fuzz targets, and this one is specifically designed to test make credential behavior, so I like it!

I'm also curious about the 0% though. OSS-Fuzz did find logic bugs (#368 was one result), so maybe can you explain a bit more? +1 to @ia0's questions.

**Comment by hcyang-google** [comment]
- The PR description mentions "0% probability with random inputs", but fuzzing is not random, it's coverage-guided. For example, fuzzers are able to reconstruct file formats like PDF without help.
- The PR description mentions coverage. How was this measured? And on which corpus? I'd suggest using the existing OSS-Fuzz corpus or building one with 24h to 7d of continuous sequential fuzzing.
- One way to help fuzzing is to provide examples or a starting corpus.

Answering these together. As fuzzer is coverage-guided, if you fuzz it long enough it will start to find interesting input and mutate more of those. But (arguably) if our fuzzer reaches interesting paths with pure random bytes, then it can have decent coverage in minutes without starting corpus, and might be more efficient on OSS-Fuzz too. This conclusion might not be correct, but the practice in our team is that we used to locally-run our new fuzzer for minutes and see if it has decent coverage before submitting that fuzzer. If not we fix the fuzzer logic so that it reaches interesting paths more easily. So seems that we have different ways of doing this, it's good that we can learn from your practices if OSS-fuzzing is efficient enough even without fine-tuning fuzzer's behavior. On the other hand making fuzzer reach interesting paths with pure random bytes seems to not have obvious disadvantages too; if it's efficient in minutes it should be efficient in days too.
I measured both cases' coverage with empty corpus and minutes of running on my workstation.

- The problem of converting the fuzzer input to valid input only a fraction of the time means the fuzzer input is not homogeneous and thus can't be easily reused/mutated. A better approach would be to have 2 fuzzing targets: one using the fuzzer input as raw input, and one using the fuzzer input to generate valid input. In some way, the first version is more end-to-end while the second is more like fuzzing the internal functions (we could actually fuzz them directly to avoid unnecessary processing which could accidentally restrict coverage).

I never thought of this before. So choosing type A of input with X% and type B of input with Y%... limits the fuzzers ability significantly? If that's the case I agree we should probably add a target. But the logic of this CL is mainly that this target is for MakeCredential, and I want it to reach the internals of MakeCredential command processing more easily.

- Regarding validity, there's also the questions of mutation. Do we want to slightly mutate the input such that it may also be invalid although it looks valid (e.g. reorder items in a map).

I'm thinking of letting the non-valid (original) path to discover all the cases where the input isn't valid. I'm not sure if we want to add more specific logic to manipulate input to reach possibly interesting paths. Now the logic is that one case is arbitrary inputs, and another is arbitrary valid inputs.

File: src/ctap/data_formats_fuzz.rs
Comment (hcyang-google) [reply src/ctap/data_formats_fuzz.rs]: ArbitraryCbor is used on types which are not one-to-one corresponding to cbors. For example `MakeCredentialOptions` has two fields but is parsed from a cbor map with three. Also, some fields of the types have defaults in case the map doesn't contain that entry, but some fields are required to exist. If we generate arbitrary types and converting them back to CBOR, we miss some valid CBOR map structures that the type can be parsed from.

**Comment by ia0** [comment]
Regarding the "raw input" vs "structured input" discussion, here is how I would compare both.

Advantages of raw vs structured:
- Simple (i.e. less bugs in the fuzzing code)
- 100% theoretical coverage

Advantages of structured vs raw:
- Faster effective coverage

Note that ideally we want to improve the coverage speed regardless of how much is already covered, i.e. the fuzzer is always able to regularly find new paths. This is different from both the theoretical coverage (when running with infinite time) and the initial coverage (when running for a few minutes).

[Here](https://github.com/google/fuzzing/blob/master/docs/structure-aware-fuzzing.md) is some documentation in case we want structure-aware fuzzing which could enable both faster effective coverage and 100% theoretical coverage, but to the price of being more complex. I don't think we should go this way unless we have evidence that we need it.

Regarding running the fuzzer for minutes instead of days, maybe random testing like [proptest](https://altsysrq.github.io/proptest-book/intro.html) would be more adapted for those usages. For me, the role of fuzzing is to theoretically cover all edge cases since no other testing technique permits that. We can still also use it for random testing, but that should be secondary.

Regarding splitting targets, [here](https://github.com/google/fuzzing/blob/master/docs/split-inputs.md#multiple-options) is documented that it is often the best approach. I think that if we want random testing, then it should be a separate target because the raw and structured input are very different and shouldn't be part of the same corpus.

This is still unclear to me which number you look at when comparing coverage. Is it the number in the output of the fuzzer? Or is it a separate command? I think those numbers (when comparing after at least one day of fuzzing) would help decide what we want to do.

File: src/ctap/data_formats_fuzz.rs
Comment (ia0) [reply src/ctap/data_formats_fuzz.rs]: I see, the structures we use make deserializing from CBOR non-injective. And this trait permits to theoretically generate all CBORs that would successfully deserialize to any possible struct value.

Note though that in practice this is not the case. For example for `MakeCredentialOptions`, the trait implementation always sets all map keys while it would be valid to deserialize when some keys are missing since they have a default value.

**Comment by hcyang-google** [comment]
Thanks, the information you provided is very helpful and I'll read them more thoroughly afterwards.
The only thing I want to clarify is that the structured fuzzer is not just random fuzzing, it still theoretically generates every valid input that can be parsed into MakeCredential command, and I thought the main purpose of this target is for MakeCredential internal command processing. But like you said, maybe we can directly fuzz inner functions for the command processing, so I'll see if I can skip the parsing part and directly fuzz the command processing. If that's the case, we don't need the ArbitraryCbor trait as we only need to generate the structure itself.

When comparing coverage, I used the llvm-cov command (https://llvm.org/docs/CommandGuide/llvm-cov.html). It generates a nice website view where I can see each file's function coverage and line coverage. I compared the function coverage and line coverage of the more relevant files (ctap/mod.rs, ctap/hid/mod.rs, ...).

**Comment by ia0** [comment]
> maybe we can directly fuzz inner functions for the command processing

Yes, if that's possible I think it's the best. It should be simpler and clearer.

> I used the llvm-cov command

I see. I compared 5 minutes of fuzzing with your branch with the current corpus on OSS-Fuzz and the coverage of `process_make_credential` is the same except for `large_blob_key` which takes both `Some` (3 times) and `None` (114 times) in your branch while only `None` (418 times) in OSS-Fuzz. But there is much more that is not covered in both cases, in particular stuff that depends on the persistent state (enterprise attestation, PIN hash, PIN length, UV, credential exclude list, cred protect). I don't think those would get covered with the raw vs structured input discussion. We should just setup the persistent state based on the input too (which could be raw and structured, but it's an orthogonal discussion).

I think it would also be worth having some behavior coverage, e.g. how often is `rk` true, how often is `uv` true, etc.

File: src/ctap/data_formats_fuzz.rs
Comment (hcyang-google) [reply src/ctap/data_formats_fuzz.rs]: Done.

File: src/ctap/data_formats_fuzz.rs
Comment (hcyang-google) [reply src/ctap/data_formats_fuzz.rs]: > Note though that in practice this is not the case. For example for MakeCredentialOptions, the trait implementation always sets all map keys while it would be valid to deserialize when some keys are missing since they have a default value.

What trait implementation? I only see a trait from cbor::Value to MakeCredentialOptions, not reversed. I thought commands from the host are parsed as cbor, and then parsed as MakeCredentialOptions, so we can't guarantee that no host will send a cbor map that has some missing values right?

**Comment by hcyang-google** [comment]
@ia0 fixed to only fuzz the CtapState command processing, with a separate fuzz target.

File: src/ctap/data_formats_fuzz.rs
Comment (ia0) [inline src/ctap/data_formats_fuzz.rs]: In reply to https://github.com/google/OpenSK/pull/441/files#r828940199:
> What trait implementation?

This implementation always sets all keys for the map. Ideally we want to use `Option<bool>::arbitrary()` and `cbor_option_map!`, or something equivalent.

File: src/ctap/data_formats_fuzz.rs
Comment (ia0) [reply src/ctap/data_formats_fuzz.rs]: I'm replying in a comment below in this file.

File: fuzz/fuzz_helper/src/lib.rs
Comment (ia0) [inline fuzz/fuzz_helper/src/lib.rs]: Let's be exhaustive to not forget to add cases in the future. Otherwise it's easy to forget fuzzing a new command.

File: fuzz/fuzz_helper/Cargo.toml
Comment (ia0) [inline fuzz/fuzz_helper/Cargo.toml]: Let's start consistently using double quotes for modified line.

File: fuzz/fuzz_helper/src/lib.rs
Comment (ia0) [inline fuzz/fuzz_helper/src/lib.rs]: Shouldn't the channel be arbitrary too?

**Review by ia0 — COMMENTED** [review]
I'm still not convinced we need the `ArbitraryCbor` trait (long-term). Short-term it's probably true because of how the library is designed. But already with #447 I think we could fuzz the new `CtapState::process_parsed_command` function directly and short-circuit CBOR logic.

But I think more importantly we need to decide what do we want to fuzz. As I understand the current state, we have a fuzz target for all commands taking CBOR and one fuzz target per command taking CBOR. This choice looks quite weird. I think we probably want something like:
- One or multiple fuzz targets at environment level, i.e. taking HID packets (the fuzzing code should not do any preprocessing on the fuzzer input like splitting in HID packets). Those targets are the ultimate targets since they fuzz based on user input.
- One or multiple fuzz targets at library level, i.e. taking stuff out of `Arbitrary`. Those targets are here to improve coverage. We could use their corpus to create better examples for the targets above (by serializing and splitting into packets).

What do you think?

Regarding the separate fuzz target, at first I thought we could reuse the existing one for "make credential" because it's kind of contained in target checking all commands, but the coverage of that one is actually very bad. So it's really useful to split by command. So the current split looks good to me.

File: src/ctap/data_formats_fuzz.rs
Comment (hcyang-google) [reply src/ctap/data_formats_fuzz.rs]: Ok this is a mistake, I should put option there. But might remove this trait, replying in comment below.

**Comment by hcyang-google** [comment]
Yes, with #447 I would prefer removing the trait and directly fuzz the command processing with parsed type. The interface is separated much better after it. I'll wait until it is merged to continue on this.

**Comment by ia0** [comment]
> Yes, with #447 I would prefer removing the trait and directly fuzz the command processing with parsed type. The interface is separated much better after it. I'll wait until it is merged to continue on this.

Cool, I've merged #447. Feel free to merge/rebase this PR.

File: fuzz/fuzz_helper/src/lib.rs
Comment (hcyang-google) [reply fuzz/fuzz_helper/src/lib.rs]: done

File: fuzz/fuzz_helper/Cargo.toml
Comment (hcyang-google) [reply fuzz/fuzz_helper/Cargo.toml]: done

File: fuzz/fuzz_helper/src/lib.rs
Comment (hcyang-google) [reply fuzz/fuzz_helper/src/lib.rs]: done

File: src/ctap/mod.rs
Comment (hcyang-google) [inline src/ctap/mod.rs:536]: If we don't do this, we have to use this [crate](https://docs.rs/visibility/latest/visibility/attr.make.html) or make something similar ourselves to achieve conditional visibility. We might want to eventually do it otherwise we'll run into this problem again when fuzzing other private functions to increase coverage.

**Comment by hcyang-google** [comment]
@ia0 can you check again if everything is fine? My only concern is how to deal with the visibility, which I put in the comment above.

File: fuzz/fuzz_targets/fuzz_target_process_ctap2_make_credential_structured.rs
Comment (ia0) [inline fuzz/fuzz_targets/fuzz_target_process_ctap2_make_credential_structured.rs]: We can update this comment now.

File: src/ctap/mod.rs
Comment (ia0) [reply src/ctap/mod.rs:536]: I don't think this is an issue. We already control the visibility of the whole ctap module here: https://github.com/google/OpenSK/blob/c3e1b5df5058e9e85e74b69262235c43d6ac5b0f/src/lib.rs#L49-L53.

**Review by ia0 — DISMISSED** [review]
LGTM modulo the documentation. I think we can submit this and check the coverage of the OSS-Fuzz corpus on this new target in a week. In the meantime do you plan to add the same for other targets? I'm actually curious how a single structured target for all commands would perform regarding coverage. Maybe that's actually enough.

File: src/ctap/mod.rs
Comment (hcyang-google) [reply src/ctap/mod.rs:536]: Oh so ctap module is private except when testing, thus we can make public all methods we want to test. Thanks.

**Comment by hcyang-google** [comment]
Fixed the comment. I'll add other fuzz targets in an upcoming PR.

**Review by ia0 — APPROVED** [review]

**Pull request merged**
